File: solutions/lua/bank-account/1/bank-account.lua
local BankAccount = { _balance = 0 }

function BankAccount:new()
  local o = setmetatable({}, self)
  self.__index = self
  return o
end

function BankAccount:close()
  self._balance = -1
end

function BankAccount:withdraw(balance)
  if balance <= 0 or balance > self._balance then error("invalid balance") end
  self._balance = self._balance - balance
end

function BankAccount:deposit(balance)
  if balance <= 0 or self._balance < 0 then error("invalid balance") end
  self._balance = self._balance + balance
end

function BankAccount:balance()
  return self._balance
end

return BankAccount

File: solutions/ruby/bank-account/1/bank_account.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# class BankAccount
class BankAccount
  def initialize(balance = 0) = (@balance = balance
                                 @open = false)

  def balance = (raise ArgumentError unless @open

                 @balance)

  def open = (raise ArgumentError if @open

              @open = true)

  def close = (raise ArgumentError unless @open

               initialize)

  def deposit(amt) = (raise ArgumentError if !@open || amt.negative?

                      @balance += amt)

  def withdraw(amt) = (raise ArgumentError if !@open || amt.negative? || (@balance - amt).negative?

                       @balance -= amt)
end
